lib/resources/resource.go

```diff
@@ -0,0 +1,152 @@
+package resources
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+
+	yaml "gopkg.in/yaml.v2"
+
+	"github.com/kosyfrances/rundeck-zabbix/lib/zabbix"
+)
+
+// Zabbix struct holds API info and methods
+type Zabbix struct {
+	API zabbix.API
+}
+
+// Resource struct holds Rundeck resource document mapping info
+type Resource struct {
+	Name        string `json:"host" yaml:"hostname"`
+	Node        string `json:"name" yaml:"nodename"`
+	Description string `json:"description" yaml:"description"`
+	OSArch      string `yaml:"osArch"`
+	OSFamily    string `yaml:"osFamily"`
+	OSName      string `yaml:"osName"`
+	OSVersion   string `yaml:"osVersion"`
+	Tags        string `yaml:"tags"`
+	Username    string `yaml:"username"`
+	SSHKeypath  string `yaml:"ssh-keypath"`
+}
+
+// Result struct holds Zabbix hosts info
+type Result []struct {
+	HostID      string `json:"hostid"`
+	Host        string `json:"host"`
+	Name        string `json:"name"`
+	Description string `json:"description"`
+	Triggers    []struct {
+		Triggerid   string `json:"triggerid,omitempty"`
+		Description string `json:"description,omitempty"`
+		Status      string `json:"status"`
+	} `json:"triggers"`
+}
+
+func (z *Zabbix) getInfo(apiKey, URL string) (Result, error) {
+	api := z.API
+	api.Key = apiKey
+	api.URL = URL
+
+	params := struct {
+		Output         []string `json:"output"`
+		SelectTriggers []string `json:"selectTriggers"`
+	}{
+		Output:         []string{"host", "name", "description"},
+		SelectTriggers: []string{"description", "status"},
+	}
+
+	payload := api.BuildPayload(params, "host.get")
+
+	resp, err := api.MakeRequest(payload)
+	if err != nil {
+		return nil, fmt.Errorf("cannot make API request. error: %v", err)
+	}
+
+	var r struct {
+		Result `json:"result"`
+	}
+
+	err = json.NewDecoder(resp.Body).Decode(&r)
+	if err != nil {
+		return nil, fmt.Errorf("cannot decode response. error: %v", err)
+	}
+
+	return r.Result, nil
+}
+
+// MakeResource generates Rundeck resource model document in Yaml format
+// with hosts' information from Zabbix
+func (z *Zabbix) MakeResource(apiKey, URL, file string) error {
+	info, err := z.getInfo(apiKey, URL)
+	if err != nil {
+		return err
+	}
+
+	var r Resource
+
+	for _, v := range info {
+		// convert item to json
+		jsonString, err := json.Marshal(v)
```

Comment:
Hmm ... `v` is marshaled here only to unmarshaled again in `toYaml` which looks like the sole purpose of the r var defined on L85 above. Is that because of the function signature of `toYaml`?

If it's not used anywhere else you can create a resource instance there and convert to yaml directly or maybe pass v down to the function directly. Either way, you skip the marshal/unmarshal overhead